internal/format/first_pass_visitor.go

```diff
@@ -0,0 +1,187 @@
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package format
+
+import (
+	"sort"
+	"strings"
+
+	"github.com/emicklei/proto"
+	"github.com/uber/prototool/internal/settings"
+	"github.com/uber/prototool/internal/text"
+)
+
+type firstPassVisitor struct {
+	*baseVisitor
+
+	Syntax                *proto.Syntax
+	Package               *proto.Package
+	Options               []*proto.Option
+	ProbablyCustomOptions []*proto.Option
+	Imports               []*proto.Import
+
+	haveHitNonComment bool
+}
+
+func newFirstPassVisitor(config settings.Config) *firstPassVisitor {
+	return &firstPassVisitor{baseVisitor: newBaseVisitor(config.Format.Indent)}
+}
+
+func (v *firstPassVisitor) Do() []*text.Failure {
+	if v.Syntax != nil {
+		v.PComment(v.Syntax.Comment)
+		if v.Syntax.Comment != nil {
+			// special case
+			v.P()
+		}
+		v.PWithInlineComment(v.Syntax.InlineComment, `syntax = "`, v.Syntax.Value, `";`)
+		v.P()
+	}
+	if v.Package != nil {
+		v.PComment(v.Package.Comment)
+		v.PWithInlineComment(v.Package.InlineComment, `package `, v.Package.Name, `;`)
+		v.P()
+	}
+	if len(v.Options) > 0 || len(v.ProbablyCustomOptions) > 0 {
+		v.POptions(false, v.Options...)
+		v.POptions(false, v.ProbablyCustomOptions...)
+		v.P()
+	}
+	if len(v.Imports) > 0 {
+		v.PImports(v.Imports)
+		v.P()
+	}
+	return v.Failures
+}
+
+func (v *firstPassVisitor) VisitMessage(element *proto.Message) {
+	v.haveHitNonComment = true
+}
+
+func (v *firstPassVisitor) VisitService(element *proto.Service) {
+	v.haveHitNonComment = true
+}
+
+func (v *firstPassVisitor) VisitSyntax(element *proto.Syntax) {
+	v.haveHitNonComment = true
+	if v.Syntax != nil {
+		v.AddFailure(element.Position, "duplicate syntax specified")
+		return
+	}
+	v.Syntax = element
+}
+
+func (v *firstPassVisitor) VisitPackage(element *proto.Package) {
+	v.haveHitNonComment = true
+	if v.Package != nil {
+		v.AddFailure(element.Position, "duplicate package specified")
+		return
+	}
+	v.Package = element
+}
+
+func (v *firstPassVisitor) VisitOption(element *proto.Option) {
+	// this will only hit file options since we don't do any
+	// visiting of children in this visitor
+	v.haveHitNonComment = true
+	if isProbablyCustomOption(element) {
+		v.ProbablyCustomOptions = append(v.ProbablyCustomOptions, element)
+	} else {
+		v.Options = append(v.Options, element)
+	}
+}
+
+func (v *firstPassVisitor) VisitImport(element *proto.Import) {
+	v.haveHitNonComment = true
+	v.Imports = append(v.Imports, element)
+}
+
+func (v *firstPassVisitor) VisitNormalField(element *proto.NormalField) {
+	v.haveHitNonComment = true
+}
+
+func (v *firstPassVisitor) VisitEnumField(element *proto.EnumField) {
+	v.haveHitNonComment = true
+}
+
+func (v *firstPassVisitor) VisitEnum(element *proto.Enum) {
+	v.haveHitNonComment = true
+}
+
+func (v *firstPassVisitor) VisitComment(element *proto.Comment) {
+	if !v.haveHitNonComment {
```

Comment:
Ah, so does this mean the first pass visitor does not concern itself with all of the comments in the file (unless it's the first thing it hits)? And then the middleVisitor picks it up from there? 

Could we leave some descriptive comments in the type declaration for what these types concern themselves with when passing over and formatting the protobuf file?

Reply:
yeah this would be super helpful, it wasn't clear why we needed these initially

Reply:
This means the only file-level comments that are printed before syntax, file-level options, imports, package are ones at the top of the file. Commenting.